query/stdlib/influxdata/influxdb/rules.go

```diff
@@ -764,3 +767,101 @@ func (PushDownWindowAggregateRule) Rewrite(ctx context.Context, pn plan.Node) (p
+// ReadGroupPhys |> { min, max, mean, count, sum }
+//
+type PushDownGroupAggregateRule struct{}
+
+func (PushDownGroupAggregateRule) Name() string {
+	return "PushDownWindowAggregateRule"
+}
+
+func (rule PushDownGroupAggregateRule) Pattern() plan.Pattern {
+	return plan.OneOf(
+		[]plan.ProcedureKind{
+			universe.MinKind,
+			universe.MaxKind,
+			universe.MeanKind,
+			universe.CountKind,
+			universe.SumKind,
+		},
+		plan.Pat(ReadGroupPhysKind))
+}
+
+func (PushDownGroupAggregateRule) Rewrite(ctx context.Context, pn plan.Node) (plan.Node, bool, error) {
+	// Check the aggregate function spec. Require operation on _value. There
+	// are two feature flags covering count and sum. All the other cases
+	// are specifically turned off.
+	var aggregateMethod string
+	fnNode := pn
+	switch fnNode.Kind() {
+	case universe.MinKind:
+		//	if !feature.PushDownSumAggregateRest().Enabled(ctx) || !caps.HaveMin() {
+		//		return pn, false, nil
+		//	}
+
+		minSpec := fnNode.ProcedureSpec().(*universe.MinProcedureSpec)
+		if minSpec.Column != execute.DefaultValueColLabel {
+			return pn, false, nil
+		}
+		// Not enabled.
+		return pn, false, nil
+	case universe.MaxKind:
+		//	if !feature.PushDownSumAggregateRest().Enabled(ctx) || !caps.HaveMax() {
+		//		return pn, false, nil
+		//	}
+
+		maxSpec := fnNode.ProcedureSpec().(*universe.MaxProcedureSpec)
+		if maxSpec.Column != execute.DefaultValueColLabel {
+			return pn, false, nil
+		}
+		// Not enabled.
+		return pn, false, nil
+	case universe.MeanKind:
+		//	if !feature.PushDownSumAggregateRest().Enabled(ctx) || !caps.HaveMean() {
+		//		return pn, false, nil
+		//	}
+
+		meanSpec := fnNode.ProcedureSpec().(*universe.MeanProcedureSpec)
+		if len(meanSpec.Columns) != 1 || meanSpec.Columns[0] != execute.DefaultValueColLabel {
+			return pn, false, nil
+		}
+		// Not enabled.
+		return pn, false, nil
+	case universe.CountKind:
+		if !feature.PushDownGroupAggregateCount().Enabled(ctx) {
+			return pn, false, nil
+		}
+
+		countSpec := fnNode.ProcedureSpec().(*universe.CountProcedureSpec)
+		if len(countSpec.Columns) != 1 || countSpec.Columns[0] != execute.DefaultValueColLabel {
+			return pn, false, nil
+		}
+		aggregateMethod = "COUNT"
+	case universe.SumKind:
+		if !feature.PushDownGroupAggregateSum().Enabled(ctx) {
+			return pn, false, nil
+		}
+
+		sumSpec := fnNode.ProcedureSpec().(*universe.SumProcedureSpec)
+		if len(sumSpec.Columns) != 1 || sumSpec.Columns[0] != execute.DefaultValueColLabel {
+			return pn, false, nil
+		}
+		aggregateMethod = "SUM"
+	}
+
+	groupNode := fnNode.Predecessors()[0]
+	groupSpec := groupNode.ProcedureSpec().(*ReadGroupPhysSpec)
+
+	// Group spec must not already have an aggregate method.
+	if len(groupSpec.AggregateMethod) > 0 {
+		return pn, false, nil
+	}
+
+	// Rule passes.
+	rewrite := *groupSpec.Copy().(*ReadGroupPhysSpec)
+	rewrite.AggregateMethod = aggregateMethod
+	return plan.CreatePhysicalNode("ReadGroupAggregate", &rewrite), true, nil
```

Comment:
There is no new ReadGroupAggregateNode. Just the name is wrong, fixed that.

Why is the name of the node detached from the implementation? Seems like I can put anything I want in that NodeID field and it still works. So why is it there?

Reply:
We need the aggregate to disappear, so simply replacing the spec means the aggregate stays. I see there is a merge function. Are there other properties in the existing `ReadGroup` we need to preserve?

Reply:
Also I noticed the existing ReadGroup rule will call `CreatePhysicalNode`. It doesn't use the merge. Wondering why the merge exists?

Reply:
The `Merge*` functions exist because when we are merging two adjacent nodes, updating the edges is non-trivial in the general case, since the graph is doubly-linked. When just pushing a node into a source, it's a little simpler.

- If you're merging two nodes in the middle of the graph use `MergeToPhysicalNode` (or logical as the case may be). This will preserve all the edges leading to successors and predecessors of the nodes being merged.
- `ReplaceSpec` will update the operation performed by the plan node, but preserve all the edges of the node. If you wanted change an operation in the middle of the DAG, but not change the data flow at all, you might do this.
- `CreatePhysicalNode` just creates the given plan node, with no edges. If you return it from the rule, the planner will wire it up to its successors, replacing what was there before.

The NodeID exists just as a way to help format the plan (like with Graphviz or something) so we can understand what the planner did.

Reply:
@wolffcm thanks for the explanation! I had assumed that returning a new node meant that incoming edges were moved to target the new node ... but that's not the case.

Reply:
@wolffcm Looking at the physical plan node
```
type PhysicalPlanNode struct {
    edges
    bounds
    id   NodeID
    Spec PhysicalProcedureSpec

    // The trigger spec defines how and when a transformation
    // sends its tables to downstream operators
    TriggerSpec TriggerSpec

    // The attributes required from inputs to this node
    RequiredAttrs []PhysicalAttributes

    // The attributes provided to consumers of this node's output
    OutputAttrs PhysicalAttributes
}
```
The are several fields here other than edges and spec. Do these all come into play *AFTER* logical and physical planning? I noticed in the rewrites we are preserving the spec and ensuring edges are correct. I wasn't even aware there are other fields in there.

Reply:
Yes, most of fields are set after all the rules have been given a chance to work and a fixed point was reached.

https://github.com/influxdata/flux/blob/master/plan/physical.go#L43-L83

- `bounds` are set after the physical plan is generated, by looking at the `range` node. Note that the time bounds for the query must be known at plan time.
- `TriggerSpec` I think is used to determine if a transformation is "narrow" or not, i.e., whether there is a 1:1 mapping of input to output tables.
- `RequiredAttrs` and `OutputAttrs` are not used currently.  The idea (not fully formed) is that for example you might need the rows or tables sorted in a particular order, like for more efficient grouping. Those fields could be used to determine what order a node provides or requires, etc.